aws/karpenter-core
# feat: Machine Migration

Fixes # 

**Description**

This PR is the `karpenter-core` switchover to managing the cloudprovider machine lifecycle through Karpenter as mentioned in https://github.com/aws/karpenter/blob/main/designs/node-ownership.md. At a high-level this PR does the following actions:

1. Provisioning now creates a `v1alpha5.Machine` to represent a scheduling decision at the end of the provisioning loop
2. Deprovisioning now uses the `state.StateNode` representation for Candidates to handle the scheduling modeling and the deprovisioning mapping to get the Machine control-plane representation of the Node
3. Inflight Checks now reconciles on the Machine and compares the Machine and the Nodes to perform validations
4. Adds machine events to go along with Node events when the Node doesn't exist
5. Activates the `machine.Controller` to reconcile Machines with requirements and launch them at the cloudprovider
6. Activate the `informer.Machine` to reconcile Machines into the `state.Cluster`
7. Adds indexes for `node.spec.providerID` and `machine.Status.providerID`
8. Allows node finalizer termination to perform the Machine finalizer flow before fully terminating the node. This provides a compatability path so that users can still perform `kubectl delete node -l karpenter.sh/provisioner-name` but the termination flow is driven through the machine
9. Enables `ttlAfterNotRegistered` value in `karpenter-global-settings` which self-terminates a `v1alpha5.Machine` if the Machine doesn't have a corresponding Node within the ttl duration.

**How was this change tested?**

- Manual Testing of:
   - Large Scale Ups (1000 nodes)
   - Machine Linking
   - Garbage Collection
- `make e2etests`
- `make presubmit`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
